sound/soc/sof/sof-priv.h

```diff
@@ -425,6 +426,9 @@ struct snd_sof_dev {
 	unsigned int extractor_stream_tag;
 #endif
 
+	/* VirtIO fields for host and guest */
+	atomic_t reset_count;
```

Comment:
@lyakh reset_count is very generic. Can we make it a bit more specific to say what we're resetting? 
Also, why cant we use the usage_count of the SOF core device instead of this? 
Ideally, when the SOF core device suspends, the device usage will drop to 0 and then begin to increase upon resume.

Reply:
@ranj063 I can call it `dsp_reset_count`. But no, it's absolutely different from a usage count. It's a monotonic counter of DSP resets. It is used to check whether the DSP has been reset since the guest last time sent it its topology.

Reply:
@lyakh and the only time the DSP get reset is when the it runtime suspends (or system suspend) isnt it and thats the usage count I'm talking about?

Reply:
@ranj063 sorry, that isn't how I understand a *usage* count. A usage count in my understanding is a counter, that counts users. In RT-PM case it might be a counter, that counts device users. For each new user it increments, when a user releases the device, it decrements. When it reaches 0 the device can be suspended.
This - once again - is a *monotonic* counter of resets. Every time the DSP resets (yes, that happens during a suspend), the counter is incremented. It is monotonic, so it is never decremented. So, no, it isn't a usage counter.

Reply:
@lyakh I got the point about this being monotnically increasing but I guess I am missing the context about how you use this in the guest. Sorry virtio is all new to me so please bear with my questions

Reply:
@ranj063 its usage is a different question. Please, have a look at https://thesofproject.github.io/latest/developer_guides/virtualization/virtualization.html#deactivation (the "Deactivation" section) and let me know if that isn't clear enough either, maybe we need to improve it too.

Reply:
> @ranj063 sorry, that isn't how I understand a _usage_ count. A usage count in my understanding is a counter, that counts users. In RT-PM case it might be a counter, that counts device users. For each new user it increments, when a user releases the device, it decrements. When it reaches 0 the device can be suspended.
> This - once again - is a _monotonic_ counter of resets. Every time the DSP resets (yes, that happens during a suspend), the counter is incremented. It is monotonic, so it is never decremented. So, no, it isn't a usage counter.

why not decrement it on suspend? Sorry, I don't get why this needs to be monotonic. I read that documentation and obviously missed something.

Even if it was needed, why not be safe and use  atomic_long_t type. Did you do the math on when the counter overflows?

Reply:
@lyakh ok so now I understand what you're trying here. Let me summarise just to ask a couple more question:
If the DSP is suspended and gets results, the guest needs to know if it needs to know if it needs to send the topology IPCs again or not. We cannot use the usage counter because it will not indicate when the reset happened. 

1. So will the SOF_IPC_PM_VFE_POWER_STATUS response contain the current dsp_reset_count? And the guest compares it against its local count to determine if the topology IPCs need to be sent again?
2. In the case of the SOF driver, when the topology is parsed, we save multiple lists (widgets, dai, controls etc). Do the gues tplg components also get saved?
or does the guest directly only send the IPCs to the DSP without the host driver saving nothing about the guest tplg?

Reply:
@lyakh I also have an alternate suggestion: Can we modify the first_boot field to maybe call it dsp_boot_count and use that instead of adding a new field?

Reply:
@ranj063 to your questions:
1. yes, and yes
2. the guest is the same exact SOF driver, it handles its topology exactly the same way as the host, so, yes, it only requests it once and saves it afterwards. As for the host SOF driver - it isn't saving guest topologies, no. It saves some information about guests, but not complete topologies.

May I also remind, that here we're not discussing the actual VirtIO implementation, because that is about to change. There are just preparatory patches that we think will be useful in their present form with the modified VirtIO implementation too.

As for using the `first_boot` - I think that should work, thanks! But it'll have to become of type `atomic_t` then.

Reply:
@lyakh great, thanks for the clarifications!

Reply:
Oh while I'm at it. Why do we need an IPC for this? Is IPC also the only way the guest and host commincate with each other?

Reply:
> Oh while I'm at it. Why do we need an IPC for this? Is IPC also the only way the guest and host communicate with each other?

The only suitable one, yes. We define 3 communication channels between the host and the guest: IPC, audio data and position updates. That means only IPC is suitable for "command and control" flows.